### PR TITLE
Injectable naming strategy and extentable services.

### PR DESCRIPTION
**Pull request by andreicio**
Made the naming strategy injectable.
Changed property and method visibility from private to protected where there's a reasonable use case for extending the service.
Added request body to fingerprint hash.

### CODE REVIEW

File: tests/Fixtures/mocks/mock_response.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-HTTP/1.1 301 Found
+HTTP/1.1 302 Moved
```
Comment (csarrazi) [inline tests/Fixtures/mocks/mock_response.txt]: Any reason for changing the mock?

File: tests/Fixtures/mocks/GET____google.com_-____9e9a30563f33dce7531caf0b1e14f6c4.txt [hunk 2]
```diff
@@ -1,14 +0,0 @@
-HTTP/1.1 302 Found
```
Comment (csarrazi) [inline tests/Fixtures/mocks/GET____google.com_-____9e9a30563f33dce7531caf0b1e14f6c4.txt]: Likewise, any reason for deleting the file?

File: tests/Adapter/PsrAdapterTest.php [hunk 3]
```diff
@@ -11,93 +11,142 @@
...
+     */
+    protected $cacheItemPoolMock;
+    /**
+     * @var NamingStrategyInterface|PHPUnit_Framework_MockObject_MockObject
+     */
+    protected $namingStrategyMock;
+    /**
+     * @var DoctrineAdapter
+     */
+    protected $sut;
```
Comment (csarrazi) [inline tests/Adapter/PsrAdapterTest.php]: Not a good variable name. Change visibility

File: tests/Adapter/PsrAdapterTest.php [hunk 4]
```diff
@@ -11,93 +11,142 @@
...
 use Psr\Http\Message\ResponseInterface;
 
 class PsrAdapterTest extends \PHPUnit_Framework_TestCase
 {
-    public function testConstructor()
+    const TTL = 10;
+    /**
+     * @var CacheItemPoolInterface|PHPUnit_Framework_MockObject_MockObject
+     */
+    protected $cacheItemPoolMock;
```
Comment (csarrazi) [inline tests/Adapter/PsrAdapterTest.php]:
```suggestion
    private $cacheItemPoolMock;
```

File: tests/Adapter/PsrAdapterTest.php [hunk 5]
```diff
@@ -11,93 +11,142 @@
...
+    /**
+     * @var NamingStrategyInterface|PHPUnit_Framework_MockObject_MockObject
+     */
+    protected $namingStrategyMock;
```
Comment (csarrazi) [inline tests/Adapter/PsrAdapterTest.php]:
```suggestion
    private $namingStrategyMock;
```

File: tests/Adapter/PsrAdapterTest.php (hunk 3 above)
Comment (csarrazi) [inline tests/Adapter/PsrAdapterTest.php]:
```suggestion
    private $adapter;
```

File: tests/Adapter/PsrAdapterTest.php [hunk 6]
```diff
@@ -11,93 +11,142 @@
...
+    protected $sut;
+
+    protected function setUp()
     {
-        $cache = $this->getMock(CacheItemPoolInterface::class);
-        new PsrAdapter($cache, 0);
+        parent::setUp();
+        $this->cacheItemPoolMock = $this->getMock(CacheItemPoolInterface::class);
+        $this->namingStrategyMock = $this->getMock(NamingStrategyInterface::class);
+        $this->sut = new PsrAdapter($this->cacheItemPoolMock, $this->namingStrategyMock, self::TTL);
```
Comment (csarrazi) [inline tests/Adapter/PsrAdapterTest.php]:
```suggestion
        $this->adapter = new PsrAdapter($this->cacheItemPoolMock, $this->namingStrategyMock, self::TTL);
```

File: tests/Adapter/DoctrineAdapterTest.php [hunk 7]
```diff
@@ -12,36 +12,56 @@
...
+     */
+    protected $cacheMock;
+    /**
+     * @var NamingStrategyInterface|PHPUnit_Framework_MockObject_MockObject
+     */
+    protected $namingStrategyMock;
+    /**
+     * @var DoctrineAdapter
+     */
+    protected $sut;
```
Comment (csarrazi) [inline tests/Adapter/DoctrineAdapterTest.php]:
```suggestion
    private $adapter;
```

File: tests/Adapter/DoctrineAdapterTest.php [hunk 8]
```diff
@@ -12,36 +12,56 @@
...
 
-    public function testConstructor()
+    protected function setUp()
     {
-        $cache = $this->getMock('Doctrine\Common\Cache\Cache');
-        new $this->class($cache, 0);
+        parent::setUp();
+        $this->cacheMock = $this->getMock(Cache::class);
+        $this->namingStrategyMock = $this->getMock(NamingStrategyInterface::class);
+        $this->sut = new DoctrineAdapter($this->cacheMock, $this->namingStrategyMock, self::TTL);
```
Comment (csarrazi) [inline tests/Adapter/DoctrineAdapterTest.php]:
```suggestion
        $this->adapter = new DoctrineAdapter($this->cacheMock, $this->namingStrategyMock, self::TTL);
```

File: tests/Adapter/DoctrineAdapterTest.php [hunk 9]
```diff
@@ -51,34 +71,64 @@ public function testFetch()
                 'reason' => 'OK',
             ])
         ;
-        $adapter = new $this->class($cache, 0);
+        $result = $this->sut->fetch($request);
```
Comment (csarrazi) [inline tests/Adapter/DoctrineAdapterTest.php]:
```suggestion
        $result = $this->adapter->fetch($request);
```

File: tests/Adapter/DoctrineAdapterTest.php [hunk 10]
```diff
@@ -51,34 +71,64 @@ public function testFetch()
...
+            ->expects($this->once())
+            ->method('contains')
+            ->with($filename)
+            ->willReturn(false)
+        ;
+        $this->cacheMock
+            ->expects($this->never())
+            ->method('fetch');
+
+        $result = $this->sut->fetch($request);
```
Comment (csarrazi) [inline tests/Adapter/DoctrineAdapterTest.php]:
```suggestion
        $adapter = $this->sut->fetch($request);
```

File: tests/Adapter/DoctrineAdapterTest.php [hunk 11]
```diff
@@ -51,34 +71,64 @@ public function testFetch()
...
                     'version' => '1.1',
                     'reason' => 'OK',
                 ],
-                10
+                self::TTL
             );
-        $adapter = new $this->class($cache, 10);
-        $adapter->save($this->getRequestMock(), new Response(200, [], 'Hello World'));
+
+        $this->sut->save($request, new Response(200, [], 'Hello World'));
```
Comment (csarrazi) [inline tests/Adapter/DoctrineAdapterTest.php]:
```suggestion
        $this->adapter->save($request, new Response(200, [], 'Hello World'));
```

File: tests/Adapter/PsrAdapterTest.php [hunk 12]
```diff
@@ -11,93 +11,142 @@
...
-        $cache
-            ->expects($this->at(1))
-            ->method('save')
-            ->with($item)
-        ;
-        $adapter = new PsrAdapter($cache, 10);
-        $adapter->save($this->getRequestMock(), new Response(200, [], 'Hello World'));
+            ->willReturn($cacheItem);
+
+        $this->sut->save($request, new Response(200, [], 'Hello World'));
```
Comment (csarrazi) [inline tests/Adapter/PsrAdapterTest.php]:
```suggestion
        $this->adapter->save($request, new Response(200, [], 'Hello World'));
```

File: tests/Adapter/PsrAdapterTest.php [hunk 13]
```diff
@@ -11,93 +11,142 @@
...
+            ->with($request)
+            ->willReturn($filename);
+
+        $this->cacheItemPoolMock
+            ->expects($this->once())
+            ->method('getItem')
+            ->with($filename)
+            ->willReturn($cacheItem);
+
+        $result = $this->sut->fetch($request);
```
Comment (csarrazi) [inline tests/Adapter/PsrAdapterTest.php]:
```suggestion
        $result = $this->adapter->fetch($request);
```

File: tests/Adapter/PsrAdapterTest.php [hunk 14]
```diff
@@ -11,93 +11,142 @@
...
+        $this->cacheItemPoolMock
+            ->expects($this->once())
             ->method('getItem')
-            ->willReturn($item)
-        ;
-        $adapter = new PsrAdapter($cache, 0);
+            ->with($filename)
+            ->willReturn($cacheItem);
 
+        $result = $this->sut->fetch($request);
```
Comment (csarrazi) [inline tests/Adapter/PsrAdapterTest.php]:
```suggestion
        $result = $this->adapter->fetch($request);
```

File: tests/Adapter/DoctrineAdapterTest.php [hunk 15]
```diff
@@ -12,36 +12,56 @@
...
 use Psr\Http\Message\ResponseInterface;
 
 class DoctrineAdapterTest extends \PHPUnit_Framework_TestCase
 {
-    protected $class = DoctrineAdapter::class;
+    const TTL = 10;
+    /**
+     * @var Cache|PHPUnit_Framework_MockObject_MockObject
+     */
+    protected $cacheMock;
```
Comment (csarrazi) [inline tests/Adapter/DoctrineAdapterTest.php]:
```suggestion
    private $cacheMock;
```

File: tests/Adapter/DoctrineAdapterTest.php [hunk 16]
```diff
@@ -12,36 +12,56 @@
...
+    /**
+     * @var NamingStrategyInterface|PHPUnit_Framework_MockObject_MockObject
+     */
+    protected $namingStrategyMock;
```
Comment (csarrazi) [inline tests/Adapter/DoctrineAdapterTest.php]:
```suggestion
    private $namingStrategyMock;
```

File: src/Adapter/PsrAdapter.php [hunk 17]
```diff
@@ -11,22 +11,29 @@
...
+   *
+   * @param CacheItemPoolInterface $cache
+   * @param NamingStrategyInterface $namingStrategy
+   * @param int $ttl
+   */
+    public function __construct(CacheItemPoolInterface $cache, NamingStrategyInterface $namingStrategy, $ttl = 0)
     {
         $this->cache = $cache;
-        $this->namingStrategy = new HashNamingStrategy();
+        $this->namingStrategy = $namingStrategy;
```
Comment (csarrazi) [inline src/Adapter/PsrAdapter.php]: BC break

File: src/Adapter/PsrAdapter.php [hunk 18]
```diff
@@ -11,22 +11,29 @@
...
 
-    public function __construct(CacheItemPoolInterface $cache, $ttl = 0)
+  /**
+   * PsrAdapter constructor.
+   *
+   * @param CacheItemPoolInterface $cache
+   * @param NamingStrategyInterface $namingStrategy
+   * @param int $ttl
+   */
+    public function __construct(CacheItemPoolInterface $cache, NamingStrategyInterface $namingStrategy, $ttl = 0)
```
Comment (csarrazi) [inline src/Adapter/PsrAdapter.php]: BC break

File: src/Adapter/MockStorageAdapter.php [hunk 19]
```diff
@@ -11,38 +11,36 @@
...
-     * @param array  $requestHeadersBlacklist
-     * @param array  $responseHeadersBlacklist
-     */
-    public function __construct($storagePath, array $requestHeadersBlacklist = [], array $responseHeadersBlacklist = [])
+  /**
+   * @param string $storagePath
+   * @param NamingStrategyInterface $namingStrategy
+   * @param array $responseHeadersBlacklist
+   */
+    public function __construct($storagePath, NamingStrategyInterface $namingStrategy, array $responseHeadersBlacklist = [])
```
Comment (csarrazi) [inline src/Adapter/MockStorageAdapter.php]: BC break

File: src/Adapter/MockStorageAdapter.php [hunk 20]
```diff
@@ -11,38 +11,36 @@
...
     {
         $this->storagePath = $storagePath;
 
-        $this->namingStrategies[] = new SubfolderNamingStrategy($requestHeadersBlacklist);
-        $this->namingStrategies[] = new LegacyNamingStrategy(true, $requestHeadersBlacklist);
-        $this->namingStrategies[] = new LegacyNamingStrategy(false, $requestHeadersBlacklist);
+        $this->namingStrategy = $namingStrategy;
```
Comment (csarrazi) [inline src/Adapter/MockStorageAdapter.php]: BC break

File: src/Adapter/MockStorageAdapter.php [hunk 21]
```diff
@@ -54,10 +52,8 @@ public function __construct($storagePath, array $requestHeadersBlacklist = [], a
      */
     public function fetch(RequestInterface $request)
     {
-        foreach ($this->namingStrategies as $strategy) {
-            if (file_exists($filename = $this->getFilename($strategy->filename($request)))) {
-                return Psr7\parse_response(file_get_contents($filename));
-            }
+        if (file_exists($filename = $this->getFilename($this->namingStrategy->filename($request)))) {
```
Comment (csarrazi) [inline src/Adapter/MockStorageAdapter.php]: BC break, as noted before

**Review by csarrazi — CHANGES_REQUESTED** [review]
Thank you @andreicio for your PR. However, there are a few things which need to be addressed:

1. Please avoid any breaking changes. These include:
    - Changing the constructor arguments for the adapters
    - Removing instantiation of the naming strategies in the constructors (otherwise these have to be constructed somewhere else)
2. Please split this PR in two distinct PRs:
    - One for enabling naming strategy injection
    - One for providing extension points for some classes (which need to be discussed). Note that the current implementation is by design, and I would rather have people extending the bundle through composition rather than inheritance (I even considered using final classes at some point).

**Comment by andreicio** [comment]
> Thank you @andreicio for your PR. However, there are a few things which need to be addressed:
> 
> 1. Please avoid any breaking changes. These include:

I was thinking of going for a new version of the bundle (maybe even remove the deprecated LegacyNamingStrategy?)
>    
>    * Changing the constructor arguments for the adapters

Add a setter method for the namingStrategy service that would be added as a call in the service definition? Or just add another optional constructor argument at the end?
>    * Removing instantiation of the naming strategies in the constructors (otherwise these have to be constructed somewhere else)

I noticed the service definitions are in the main guzzle bundle, so I was thinking of extending the mock plugin config to allow selecting the prefered storage adapter and naming strategy. I know this would not be BC, but, as I mentioned, I was thinking of a version change.
> 2. Please split this PR in two distinct PRs:
>    
>    * One for enabling naming strategy injection
>    * One for providing extension points for some classes (which need to be discussed). Note that the current implementation is by design, and I would rather have people extending the bundle through composition rather than inheritance (I even considered using final classes at some point).

Ok, I'll revert changes that make the storage adapters extendable. I would keep the protected visibility for the blacklist property in the AbstractNamingStrategy class though.

So, to summarize, what is your prefered solution for enabling naming strategy injection and keeping BC? Setter method, additional optional constructor argument, or something else?

**Comment by csarrazi** [comment]
Hi @andreicio, please find my comments below.

> > Thank you @andreicio for your PR. However, there are a few things which need to be addressed:
> > 
> > 1. Please avoid any breaking changes. These include:
> 
> I was thinking of going for a new version of the bundle (maybe even remove the deprecated LegacyNamingStrategy?)

This is fine by me. However, although introducing BC, I would like to ensure that the contracts are not entirely broken. I would prefer adding a new argument to the constructor, with a default `null` value at the end of the argument list.
Second, if no naming strategy is provided, we should ensure that the current naming strategy is used. This will ensure that the library won't have any breaking change, while still providing new functionality.

Keep in mind that people not only use the bundle, but also use the various libraries as standalone dependencies (reason why the various middleware were actually split from the bundle)
> 
> > * Changing the constructor arguments for the adapters
> 
> Add a setter method for the namingStrategy service that would be added as a call in the service definition? Or just add another optional constructor argument at the end?

As mentioned before, optional constructor argument at the end.

> 
> > * Removing instantiation of the naming strategies in the constructors (otherwise these have to be constructed somewhere else)
> 
> I noticed the service definitions are in the main guzzle bundle, so I was thinking of extending the mock plugin config to allow selecting the prefered storage adapter and naming strategy. I know this would not be BC, but, as I mentioned, I was thinking of a version change.

I don't think we actually need to make the mock plugin extendable. We just need to provide a means to inject a different storage adapter and naming strategy.

> 
> > 1. Please split this PR in two distinct PRs:
> >    
> >    * One for enabling naming strategy injection
> >    * One for providing extension points for some classes (which need to be discussed). Note that the current implementation is by design, and I would rather have people extending the bundle through composition rather than inheritance (I even considered using final classes at some point).
> 
> Ok, I'll revert changes that make the storage adapters extendable. I would keep the protected visibility for the blacklist property in the AbstractNamingStrategy class though.

I don't see what adding switching the visibility to protected will actually provide, especially when you can actually inject the blacklist.

Actually, the blacklists which are already predefined in the classes should actually be removed at one point, to be migrated elsewhere (probably in the Middleware factory class)

**Comment by andreicio** [comment]
Updated version that only includes the ability to inject a naming strategy in the 3 existing adapters.

There is an additional change that might be required, to composer json and lock, since I got an error on running composer install. The proposed change is removing "minimum-stability" and regenerating the lock file by running composer update. This could be done in a separate pull request though.

**Comment by andreicio** [comment]
Have you had time to review the new version?

**Comment by csarrazi** [comment]
Whoops, sorry I didn't see that you had pushed around end of June. I'll try to take a look soon (end of the week, as these days are pretty busy)

File: src/Adapter/MockStorageAdapter.php [hunk 22]
```diff
@@ -57,6 +62,8 @@ public function fetch(RequestInterface $request)
         foreach ($this->namingStrategies as $strategy) {
             if (file_exists($filename = $this->getFilename($strategy->filename($request)))) {
                 return Psr7\parse_response(file_get_contents($filename));
+            } else {
+                var_dump($filename);
```
Comment (csarrazi) [inline src/Adapter/MockStorageAdapter.php]: Isn't that a remaining debug statement?

**Review by csarrazi — CHANGES_REQUESTED** [review]
Just removed the debug statement, and the PR will be good to merge!

Thanks for the hard work @andreicio!

File: src/Adapter/MockStorageAdapter.php (hunk 22 above)
Comment (andreicio) [reply src/Adapter/MockStorageAdapter.php]: Sorry about that! Goes to prove there's no amount of experience that protects me from noob mistakes :)

**Review by csarrazi — APPROVED** [review]

**Comment by csarrazi** [comment]
Thank you @andreicio!

**Comment by nickvanderzwet** [comment]
Is it possible to make a new release for this change?

**Comment by csarrazi** [comment]
Yes, I will do so once I have access to my laptop (in a couple of weeks) :)
